webcrypto/elliptic_curve.go

```diff
@@ -1,5 +1,214 @@
+	// Its name should be set to either ECDSA or ECDH.
+	Algorithm
+
+	// NamedCurve holds (a String) the name of the curve to use.
+	// You can use any of the following: CurveKindP256, CurveKindP384, or CurveKindP521.
+	NamedCurve EllipticCurveKind `json:"namedCurve"`
+}
+
+// NewEcKeyGenParams creates a new EcKeyGenParams instance from a goja.Value.
+func NewEcKeyGenParams(rt *goja.Runtime, v goja.Value) (EcKeyGenParams, error) {
+	if v == nil {
+		return EcKeyGenParams{}, NewError(0, SyntaxError, "algorithm is required")
+	}
+
+	var params EcKeyGenParams
+	if err := rt.ExportTo(v, &params); err != nil {
+		return EcKeyGenParams{}, NewError(0, SyntaxError, "algorithm is invalid")
+	}
+
+	if err := params.Validate(); err != nil {
+		return EcKeyGenParams{}, err
+	}
+
+	params.Normalize()
+
+	return params, nil
+}
+
+// Ensure EcKeyGenParams implements the Validator interface.
+var _ Validator = &EcKeyGenParams{}
+
+// Validate validates the EcKeyGenParams instance. It implements the Validator
+// interface.
+func (e EcKeyGenParams) Validate() error {
+	if e.Name == "" {
+		return NewError(0, SyntaxError, "algorithm name is required")
+	}
+
+	if !strings.EqualFold(e.Name, ECDSA) {
+		return NewError(0, SyntaxError, "invalid algorithm name")
+	}
+
+	if !IsEllipticCurve(string(e.NamedCurve)) {
+		return NewError(0, NotSupportedError, "invalid elliptic curve name")
+	}
+
+	return nil
+}
+
+// Ensure EcKeyGenParams implements the Normalizer interface.
+var _ Normalizer = &EcKeyGenParams{}
+
+// Normalize normalizes the algorithm name and elliptic curve name. It implements the
+// Normalizer interface.
+func (e *EcKeyGenParams) Normalize() {
+	e.Name = NormalizeAlgorithmName(e.Name)
+	e.NamedCurve = EllipticCurveKind(strings.ToUpper(string(e.NamedCurve)))
+}
+
+// Ensure EcKeyGenParams implements the CryptoKeyPairGenerator interface.
+var _ CryptoKeyPairGenerator[crypto.PrivateKey, crypto.PublicKey] = &EcKeyGenParams{}
+
+// GenerateKeyPair generates a new Elliptic Curve key pair.
+// It implements the CryptoKeyPairGenerator interface.
+//
+//nolint:funlen
+func (e EcKeyGenParams) GenerateKeyPair(
+	extractable bool,
+	keyUsages []CryptoKeyUsage,
+) (*CryptoKeyPair[crypto.PrivateKey, crypto.PublicKey], error) {
+	if e.Algorithm.Name != ECDSA && e.Algorithm.Name != ECDH {
+		return nil, NewError(0, ImplementationError, "invalid algorithm name")
+	}
+
+	// There are no implementation of the Diffie-Hellman key exchange algorithm
+	// in the Go standard library, so we protect our users from ourselves (and themselves),
+	// and refuse to generate ECDH keys until it is implement in the Go stdlib.
+	//
+	// As of November 2022, a proposal has been accepted, and ECDH implementation into
+	// the golang standard library [is in progress].
+	//
+	// [is in progress]: https://github.com/golang/go/issues/52221
+	if e.Algorithm.Name == ECDH {
+		return nil, NewError(0, NotSupportedError, "ECDH key generation is not supported, yet")
+	}
```

Comment:
Implemented in 1.20: https://pkg.go.dev/crypto/ecdh

Reply:
Indeed, I found the proposal as I was implementing, but considering that we're still with 1.18, and 1.20 and is not out yet, I'll wait for it to be out, and for k6 to use it before I activate support for ECDH 👍🏻 :bow

Reply:
Fair enough.